include/swift/AST/EvaluatorDependencies.h

```diff
@@ -0,0 +1,88 @@
+//===--- EvaluatorDependencies.h - Auto-Incremental Dependencies -*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines data structures to support the request evaluator's
+// automatic incremental dependency tracking functionality.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_AST_EVALUATOR_DEPENDENCIES_H
+#define SWIFT_AST_EVALUATOR_DEPENDENCIES_H
+
+#include "swift/AST/AttrKind.h"
+#include "swift/AST/SourceFile.h"
+#include "llvm/ADT/PointerIntPair.h"
+
+namespace swift {
+
+namespace evaluator {
+
+/// The "scope" of a dependency edge tracked by the evaluator.
+///
+/// Dependency scopes come in two flavors: cascading and private. A private
+/// edge captures dependencies discovered in contexts that are not visible to
+/// to other files. For example, a conformance to a private protocol, or the use
+/// of any names inside of a function body. A cascading edge, by contrast,
+/// captures dependencies discovered in the remaining visible contexts. These
+/// are types with at least \c internal visibility, names defined or used
+/// outside of function bodies with at least \c internal visibility, etc. A
+/// dependency that has cascading scope is so-named because upon traversing the
+/// edge, a reader such as the driver should continue transitively evaluating
+/// further dependency edges.
+///
+/// A cascading edge is always conservatively correct to produce, but it comes
+/// at the cost of increased resources spent (and possibly even wasted!) during
+/// incremental compilation. A private edge, by contrast, is more efficient for
+/// incremental compilation but it is harder to safely use.
+///
+/// To ensure that these edges are registered consistently with the correct
+/// scopes, requests that act as the source of dependency edges are required
+/// to specify a \c DependencyScope under which all evaluated sub-requests will
+/// register their dependency edges. In this way, \c DependencyScope values
+/// form a stack-like structure and are pushed and popped by the evaluator
+/// during the course of request evaluation.
+///
```

Comment:
The stack is a nice solution to the problem of connecting the code execution in the frontend to figuring out the dependency context.